### PR TITLE
Object filter types

### PR DESCRIPTION
**Pull request by PascalSenn**
Current implementation is limited to schema generation. 

I didn't know how to lazy initialize a type 
Check out TODO comments in ObjectFilterFieldDescriptor.cs

Addresses: #921

### CODE REVIEW

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                RewriteType(operationKind),
+                operation);
+        }
+
+        public IObjectFilterFieldDescriptor<TObject> AllowObject(Action<IFilterInputTypeDescriptor<TObject>> descriptor)
+        {
+            var desc = FilterInputTypeDescriptor<TObject>.New(Context, typeof(TObject));
+            descriptor.Invoke(desc);
+
+            // TODO: I don't know how to initialize this type
```
Comment (PascalSenn) [inline src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: @michaelstaib  Here

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            ObjectFilterOperationDescriptor<TObject> field =
+                CreateOperation(FilterOperationKind.Equals);
+            Filters.Add(field);
+            return this;
+        }
+
+        public IObjectFilterFieldDescriptor<TObject> AllowObject<TFilter>() where TFilter : IFilterInputTypeDescriptor<TObject>
+        {
+
+            // TODO: Do wee need to pass the generic TFilter to create Option?
```
Comment (PascalSenn) [inline src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: @michaelstaib And here

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs (hunk 1 above)
Comment (michaelstaib) [reply src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: What is the invoke doing?

Comment (michaelstaib) [reply src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: ah .... I see .... to early in the  morning ;)

**Comment by michaelstaib** [comment]
OK, id did some more work to make extending types easier....

**Comment by michaelstaib** [comment]
```csharp
        public IObjectFilterFieldDescriptor<TObject> AllowObject(
            Action<IFilterInputTypeDescriptor<TObject>> descriptor)
        {
            var type = new FilterInputType<TObject>(descriptor);
            var typeReference = new SchemaTypeReference(type);
            .... 
            return this;
        }
```

You still have now to rewrite the following part:

```csharp
            ObjectFilterOperationDescriptor<TObject> field =
                CreateOperation(FilterOperationKind.Equals);
            Filters.Add(field);
```

Do we need to add a new operation kind for that?

The `typeReference` is the return type of the operation field so this would need to become the `field.Type(typeReference)`.

File: src/Core/Types.Filters/FilterInputTypeDescriptor.cs [hunk 3]
```diff
@@ -116,6 +116,14 @@ public class FilterInputTypeDescriptor<T>
                 return true;
             }
 
+            if (property.PropertyType.IsClass)
```
Comment (PascalSenn) [inline src/Core/Types.Filters/FilterInputTypeDescriptor.cs]: Is this gate safe?

File: src/Core/Types.Filters/FilterInputTypeDescriptor.cs [hunk 4]
```diff
@@ -183,6 +191,22 @@ public class FilterInputTypeDescriptor<T>
                 nameof(propertyOrMethod));
         }
 
+        public IObjectFilterFieldDescriptor<TObject> Filter<TObject>(
+            Expression<Func<T, TObject>> propertyOrMethod) where TObject : class
```
Comment (PascalSenn) [inline src/Core/Types.Filters/FilterInputTypeDescriptor.cs]: I wonder if there is a better generic constraint than this...

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+{
+    public class ObjectFilterFieldDescriptor
+       : FilterFieldDescriptorBase
+       , IObjectFilterFieldDescriptor
+    {
+        private readonly Type type;
+        public ObjectFilterFieldDescriptor(
+            IDescriptorContext context,
+            PropertyInfo property,
+            Type type)
```
Comment (PascalSenn) [inline src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: Should i remove type and just read the type over the property?

File: src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs [hunk 6]
```diff
@@ -63,6 +63,17 @@ public void Type<TInputType>(TInputType inputType)
             Definition.Type = new SchemaTypeReference(inputType);
         }
 
+        public void Type(ITypeReference typeReference)
```
Comment (PascalSenn) [inline src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs]: Does this makes sense?

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            var operation = new FilterOperation(
+                type,
+                operationKind,
+                Definition.Property);
+
+            return ObjectFilterOperationDescriptor.New(
+                Context,
+                this,
+                CreateFieldName(operationKind),
+                new ClrTypeReference(typeof(FilterInputType<>).MakeGenericType(type), Definition.Type.Context, true, true),
```
Comment (PascalSenn) [inline src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: hm......

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor~1.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            ObjectFilterOperationDescriptor<TObject> field =
+                CreateOperation(FilterOperationKind.Equals);
+            field.Type<TFilter>();
+            Filters.Add(field);
+
+            return field;
+        }
+
+        public new IObjectFilterOperationDescriptor<TObject> AllowObject()
```
Comment (PascalSenn) [inline src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor~1.cs]: I introduced this API for implicit bindings

**Comment by PascalSenn** [comment]
@michaelstaib  Thanks, helped a lot. I could resolve the schema and added a few tests. 
I do have some open questions. I added reviews for them. Basically i highlighted the parts I'm not sure if I've done everything right 👍 

I guess after this review we can start to work on the visitor

File: src/Core/Types.Filters/FilterInputTypeDescriptor.cs (hunk 4 above)
Comment (michaelstaib) [reply src/Core/Types.Filters/FilterInputTypeDescriptor.cs]: And also what about interfaces or structs that represent a complex type.

Comment (michaelstaib) [reply src/Core/Types.Filters/FilterInputTypeDescriptor.cs]: Maybe the name is wrong.... is it really a filter? I mean in the case of an object?

Comment (michaelstaib) [reply src/Core/Types.Filters/FilterInputTypeDescriptor.cs]: And what if an object is comparable :D

File: src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs (hunk 5 above)
Comment (michaelstaib) [reply src/Core/Types.Filters/Object/ObjectFilterFieldDescriptor.cs]: I will have to check.

File: src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs (hunk 6 above)
Comment (michaelstaib) [reply src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs]: Not on public stuff .... the ITypeReference is only for internal purposes ... they basically also are placeholders for the actual type while the schema builder is creating the schema.

Comment (michaelstaib) [reply src/Core/Types/Types/Descriptors/ArgumentDescriptorBase~1.cs]: why do you need it there anyway?

**Comment by michaelstaib** [comment]
I am focusing at the moment to finish 10x tasks... after I have finished those I will focus in this one with you.

**Comment by michaelstaib** [comment]
Is there something blocking or can we get this one going again?

**Comment by PascalSenn** [comment]
I had a look in it a few days ago. 
I was trying to figure out the best way for the visitor. There is still that comment you left there for object types. I guess you already knoe exactly how to do it?

Type generation seems to work. You may want to have a closer look at the PR. There are a few things that are not very nice.

**Comment by PascalSenn** [comment]
@michaelstaib  I'm kind of stuck on the visitor.  Once this is done i could start on the tests

**Comment by michaelstaib** [comment]
I will help you tomorrow on that one. Sorry for the wait.

**Comment by michaelstaib** [comment]
@PascalSenn I have to fix  the circle/ci ... the container still uses .net core 3 preview .... 

I will fix it and hit rebuild.

**Comment by michaelstaib** [comment]
OK, pushed the build fix.

**Comment by michaelstaib** [comment]
Circle CI is green now ....

**Comment by michaelstaib** [comment]
I wait for the sonar results then I start the review.

**Comment by michaelstaib** [comment]
I have fixed some sonar issues and I think we can merge it once it is green. We will test it internally and see if it has what it takes.

If we find issues we can still iterate on it.

**Pull request merged**